### DISCUSSION TITLE
Global variables in rexx code

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#186] [Q&A]
Although my global sample runs, I am not sure that I understand everything correctly. I split the global processing in SETG(lobal) and GETG(lobal), called from main.rexx. 

```
/* rexx */
  options levelb
    call setg("Fred")
    say 'MAIN 'getg()
return

/* prototype statements */
setg: procedure = .int
   arg str =.string

getg: procedure = .string   
      arg str = ""

/* global set and get functions */ 
.globals=1
g0 .expose=my_globals

setg()      .locals=3 .expose=setg.setg
               copy g0,a1
               ret

 .globals=1
g0 .expose=my_globals

getg()     .locals=1 .expose=getg.getg
               say g0
               ret g0
```
1/ Is my assumption correct, that the global g0 is set in a "global-pool" called my_globals, therefore I can have g0 defined several times, as long it is in a different pool. You just need to address the appropriate pool, aka pool-name.variable-name. 
2/ in the main rexx there is no additional reference to global definitions necessary (if I use it the way I did it)?

**Comment by adesutherland** [comment]
1/ Yes
 The g0 .expose=my_globals is valid for the entire file - so if you have setg and getg in one file you only need to map g0 once.

In another file you can have

g0 .expose=my_globals 

which links up the two references via "my_global" (which you might call instead "globals.timestamp" or something).

You can also have 

g1 .expose=my_globals - or g{anything} - in sum, each rxas file can map the globals to specific registers as it seems fit.

 2/ Correct

**Comment by Peter-Jacob** [comment]
I can't define 2 globals in the same pool, can I?
```
g0    .expose=time_globals
g1    .expose=time_globals

Error Severity 1
3:14 - Error at "time_globals", duplicate exposed index
```
this works:
```
g0 .expose=time_global1
g1 .expose=time_global2

```
so the expose name is not a pool name it's the variable name

**Comment by adesutherland** [comment]
Yeah ... it is the variable name (global index) not a pool name, I understand you point better. This is why I suggest the "xx.bb" type format, in a way the xx is the pool and the bb is the specific global register index. And this can be mapped to any g{n}.

But also you can't expose the same global index twice from 1 file (and after all why would you need to?)

Hope this makes sense?

A

**Comment by Peter-Jacob** [comment]
Makes absolute sense. There remains one question: Do we have a way to define:
   g0 .expose=time_global1
within rexx source? I tried several things, no one worked. Assembler prefix doesn't work of course because it's not an instruction, but a definition.

**Reply by adesutherland** [reply]
No you don't

**Reply by adesutherland** [reply]
But this also means you can **only** access g0 from assembler not from rexx - so you need have the getg() low level function in assembler called by higher level rexx functions

One day we will have a way of doing globals in rexx level b - one day

**Reply by Peter-Jacob** [reply]
good to know, but not a problem. I'll proceed with my setg/getg approach.
